src/lib.rs

```diff
@@ -153,50 +153,31 @@ pub enum Error {
     #[error("Message data does not validate per the schema: {0}")]
     DataValidation(String),
 
-    /// Publisher failed to publish a message
-    #[error("Publisher failed to publish a message batch")]
-    Publisher(#[source] Box<dyn std::error::Error + Send + Sync>),
-
-    /// Publisher failed to publish multiple batches of messages
-    #[error("Publisher failed to publish multiple batches (total of {} errors)", _1.len() + 1)]
-    PublisherMultiple(
-        #[source] Box<dyn std::error::Error + Send + Sync>,
-        Vec<Box<dyn std::error::Error + Send + Sync>>,
-    ),
-}
-
-type AnyError = Box<dyn std::error::Error + Send + Sync>;
+    /// Could not validate message data
+    #[error("Could not add a message to the builder")]
+    AddMessage(#[source] Box<dyn std::error::Error + Send + Sync>),
 
-/// The special result type for [`Publisher::publish`](trait.Publisher.html)
-#[derive(Debug)]
-pub enum PublisherResult<Id> {
-    /// Publisher succeeded.
-    ///
-    /// Contains a vector of published message IDs.
-    Success(Vec<Id>),
-    /// Publisher failed to publish any of the messages.
-    OneError(AnyError, Vec<ValidatedMessage>),
-    /// Publisher failed to publish some of the messages.
-    ///
-    /// The error type has a per-message granularity.
-    PerMessage(Vec<Result<Id, (AnyError, ValidatedMessage)>>),
+    /// Could not validate message data
+    #[error("Could not publish the message")]
+    PublishOneMessage(#[source] Box<dyn std::error::Error + Send + Sync>),
 }
 
 /// Interface for message publishers
 pub trait Publisher {
     /// The list of identifiers for successfully published messages
     type MessageId: 'static;
-    /// The future that the `publish` method returns
-    type PublishFuture: Future<Output = PublisherResult<Self::MessageId>> + Send;
+    /// The message error that this publisher returns
+    type MessageError: std::error::Error + Send + Sync + 'static;
+    /// The stream of results that the `publish` method returns
+    type PublishStream: Stream<Item = Result<Self::MessageId, Self::MessageError>> + Unpin;
 
     /// Publish a batch of messages
     ///
-    /// # Return value
-    ///
-    /// Shall return [`PublisherResult::Success`](PublisherResult::Success) only if all of the
-    /// messages are successfully published. Otherwise `PublisherResult::OneError` or
-    /// `PublisherResult::PerMessage` shall be returned to indicate an error.
-    fn publish(&self, topic: &'static str, messages: Vec<ValidatedMessage>) -> Self::PublishFuture;
+    /// The output stream shall return a result for each message in `messages` slice in the same
+    /// order.
+    fn publish<'a, I>(&self, topic: &'static str, messages: I) -> Self::PublishStream
+    where
+        I: Iterator<Item = &'a ValidatedMessage> + DoubleEndedIterator + ExactSizeIterator;
```

Comment:
`ValidatedMessage`s get produced internally within hedwig after the user-constructed `Message`s are validated.

Reply:
`Stream` cannot exactly support `DoubleEndedIterator` or `ExactSizeIterator`. While the only "real" publisher does not need either of those (and it would make some sense for it to accept a `Stream`) I don’t want to make such a change in this PR. Specifically because I don’t want to spend the time right now thinking about what the consequences of such a change might end up being.

Reply:
alright, that sounds reasonable to me